X atom types enum

Danacus opened the pull request:
I had to create a new PR because I force-pushed some changes, I'm sorry for the inconvenience. I managed to solve the key input issues I was having and had to fix a few more things that were causing panics. Now everything appears to be working and it should be ready for review.

sminez commented on src/core/helpers.rs:
```diff
@@ -139,15 +139,15 @@ pub fn parse_key_binding(pattern: impl Into<String>, known_codes: &CodeMap) -> O
 
-    pub fn intern_atom(conn: &xcb::Connection, name: &str) -> Result<u32> {
-        xcb::intern_atom(conn, false, name)
+    pub fn intern_atom(conn: &xcb::Connection, name: &Atom) -> Result<u32> {
```
The purpose of this function is to intern an atom directly (which is why it makes the underlying `xcb` call rather than using an `XConn`) so it should stay as it is.

sminez commented on src/core/helpers.rs:
```diff
@@ -180,13 +180,13 @@ pub(crate) mod xcb_util {
+            &conn,                                       // xcb connection to X11
+            xcb::PROP_MODE_REPLACE as u8,                // discard current prop and replace
+            id,                                          // window to change prop on
+            intern_atom(&conn, &Atom::NetWmWindowType)?, // prop to change
```
going forward, having a way to pull atoms directly for known atoms would be good but this should be left as it was for the time being.

sminez commented on src/core/macros.rs:
```diff
@@ -82,3 +82,49 @@ macro_rules! gen_keybindings(
+ * Create a slice with `$slice_name` as name containing all enum variants  
+ */
+#[macro_export]
+macro_rules! gen_enum_with_slice {
```
If this is only used in the `xconnection` module can we move the definition there and not export it? Given that this is a hand implemented macro (as opposed to just using strum on the enums in question) it is not really something I feel happy exposing in the public API.

sminez commented on src/core/macros.rs:
```diff
@@ -82,3 +82,49 @@ macro_rules! gen_keybindings(
+        }
+
+        impl $enum_name {
+            pub(crate) fn as_str(&self) -> &str {
```
Can we do this as implementing `Into<&str>` instead? That way we make use of exisiting idioms and code that accepts an `impl Into<&str>`

sminez commented on src/core/macros.rs:
```diff
@@ -82,3 +82,49 @@ macro_rules! gen_keybindings(
+                }
+            }
+
+            pub(crate) fn from_str(name: &str) -> crate::Result<Self> {
```
This is `try_from` not `from` as it is not guaranteed to succeed. (same comment about implementing the corresponding trait as above)

sminez commented on src/core/xconnection.rs:
```diff
@@ -53,62 +53,85 @@ const EVENT_MASK: &[(u32, u32)] = &[(
-    "_NET_WM_WINDOW_TYPE_NORMAL",
-];
+gen_enum_with_slice!(
+    /**
```
Stylisitic nitpicking: if it's a single line doc comment then please use `///` over `/** ... */` to reduce line noise.

sminez commented on src/core/xconnection.rs:
```diff
@@ -314,7 +337,7 @@ pub trait XConn {
 
     /// Send an X event to the target window
-    fn send_client_event(&self, id: WinId, atom_name: &str) -> Result<()>;
+    fn send_client_event(&self, id: WinId, atom_name: &Atom) -> Result<()>;
```
The public API of `XConn` should not be changing as part of this PR: usage of hard coded strings internally can (and should) be updated, but these methods need to be able to accept arbitrary strings in order to support end users requesting atoms not currently in the Enum defined above. Penrose is not fully supporting all EWMH atoms directly yet (and may never do so) but if users wish to work with atoms directly we need to be able to support that.

Danacus replied:
Okay, that makes sense.

sminez commented on src/core/xconnection.rs:
```diff
@@ -364,13 +387,13 @@ pub trait XConn {
      * Can fail if the property name is invalid or we get a malformed response from xcb.
      */
-    fn str_prop(&self, id: u32, name: &str) -> Result<String>;
+    fn str_prop(&self, id: u32, name: &Atom) -> Result<String>;
```
same here

sminez commented on src/core/xconnection.rs:
```diff
@@ -364,13 +387,13 @@ pub trait XConn {
      * Can fail if the property name is invalid or we get a malformed response from xcb.
      */
-    fn atom_prop(&self, id: u32, name: &str) -> Result<u32>;
+    fn atom_prop(&self, id: u32, name: &Atom) -> Result<u32>;
```
and here

sminez commented on src/core/xconnection.rs:
```diff
@@ -400,20 +424,33 @@ impl XcbConnection {
                     .atom();
 
-                (*atom, val)
+                (atom.clone(), val)
```
if `Atom` implements `Copy` then this deref should still work

sminez commented on src/core/xconnection.rs:
```diff
@@ -53,62 +53,85 @@ const EVENT_MASK: &[(u32, u32)] = &[(
+    [XEmbedInfo, "_XEMBED_INFO"],
+});
+
+gen_enum_with_slice!(
```
Making these atoms a different type feels like it is going to cause some issues down the line. Atoms are frequently used as arguments to XCB API calls and some of the more generic methods may have issues with this change. Please keep them as a single type: sorry I missed the fact that you had done this in your comment on the original issue.

Danacus replied:
At first it looked like a good idea, but when I was changing other parts of the code, I realised it might not have been the best choice. So I agree that it wasn't the best idea.

sminez commented on src/core/xconnection.rs:
```diff
@@ -400,20 +424,33 @@ impl XcbConnection {
             .collect();
 
-        let auto_float_types: Vec<&str> = AUTO_FLOAT_WINDOW_TYPES.to_vec();
+        let window_types: HashMap<WindowType, u32> = WINDOW_TYPES
```
see comment above about not splitting window types out into their own enum

sminez commented on src/core/xconnection.rs:
```diff
@@ -445,20 +482,32 @@ impl XcbConnection {
     }
 
-    fn atom(&self, name: &str) -> Result<u32> {
+    fn atom(&self, atom: &Atom) -> Result<u32> {
```
If this function is now accepting a closed enum, it is no longer possible to fail (we will have errored on window manager initi when we tried to intern all of the atoms).

sminez commented on src/core/xconnection.rs:
```diff
@@ -474,13 +523,13 @@ impl XcbConnection {
+            &self.conn,                              // xcb connection to X11
+            false,                                   // should the property be deleted
+            id,                                      // target window to query
+            self.known_atom(&Atom::NetWmWindowType), // the property we want
```
this should just be a `get` from the hashmap

Danacus replied:
Does this mean that `atom` and `known_atom` are no longer needed?

sminez reviewed (changes requested):
This is a very broadly scoped diff for two commits, which also goes beyond the scope of the target issue (#41) in that it modifies the public API of the `XConn` trait. You mention having issues with key events not being picked up and panics in your PR description (in addition to saying that you now have everything sorted) but haven't provided any details on what those issues were or _how_ they were resolved.

I'm afraid I don't have time to review (or re-review) a PR like this in the future. The method proposed in the issue comments was fine: implement the new enum and remove internal use of hard coded string literals. The contents of this PR actually touch on a number of other things that are not going to be possible to merge at this time.

Danacus commented:
Thank you for taking your time to review this PR! I'm still learning both Rust and learning to collaborate to open source projects, but I understand that I could be a bit of a burden for you. If you don't have time, you don't have to feel forced to immediately respond to my PRs.

The key input issue was caused by me using `startx` incorrectly. There's a difference between `startx penrose -- :2 -keeptty` and `startx penrose -- -keeptty  :2`. When using the latter key input does not work. I also tried to use `Xephyr` to make debugging easier, but there I'm having the same issue with key input not working. This might be an issue with penrose, because other WMs work just fine. I might open an issue for that if I can't find the solution myself.

The panics were from an `unimplemented!` I had written in the `from_str` function. That's why I changed it such that it returns a `Result` instead.

Danacus commented on src/core/xconnection.rs:
```diff
@@ -752,8 +747,8 @@ impl XConn for XcbConnection {
     fn send_client_event(&self, id: WinId, atom_name: &str) -> Result<()> {
-        let atom = self.atom(atom_name)?;
-        let wm_protocols = self.known_atom("WM_PROTOCOLS");
+        let atom = *self.atoms.get(&atom_name.try_into()?).unwrap();
```
I'm not sure if you prefer using `Atom::try_from(atom_name)?` here instead.

Danacus commented on src/core/xconnection.rs:
```diff
@@ -811,117 +855,121 @@ impl XConn for XcbConnection {
 
     fn set_wm_properties(&self, workspaces: &[&str]) {
         // xcb docs: https://www.mankier.com/3/xcb_change_property
+        xcb::change_property(
```
I'm sorry if the diff looks a bit confusing here, I'm guessing it's because of the formatting changes. All I did here was replace `known_atom()` with `*self.atoms.get().unwrap()`.

Danacus commented on src/core/xconnection.rs:
```diff
@@ -469,17 +524,6 @@ impl XcbConnection {
         })
     }
 
-    fn atom(&self, name: &str) -> Result<u32> {
```
I think these are no longer needed.

Danacus commented:
I have applied the changed you requested and rebased this branch on `develop`. (I'm not sure if I should merge the upstream branch instead of rebasing).

sminez commented on src/core/xconnection.rs:
```diff
@@ -419,11 +474,11 @@ impl XcbConnection {
             .map(|atom| {
                 // false == always return the atom, even if exists already
-                let val = xcb::intern_atom(&conn, false, atom)
+                let val = xcb::intern_atom(&conn, false, (*atom).into())
```
does changing line 479 to `.map(|&atom| {` allow you to simplify this to just `atom.into()` ?

(see [this playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=42cc9817ce85dcecdb768ea9a7932375) for an example of what I mean)

Danacus replied:
That does look a bit nicer, thanks!

sminez commented:
@Danacus please can you remove your changes to functions that are expecting `&str` parameters. Those functions should be accepting arbitrary atom names from the caller and attempting to intern / use them with the XCB API directly: the enum should only be being used to ensure that hard coded uses of atoms in the codebase are for known atoms as a way of reducing human error.

Sorry if this wasn't clear in my initial review: the scope of this issue is simply to add the enum and replace existing static `&str` references with usage of the enum. Not to modify any parts of the API.

Danacus commented:
I'm guessing you're talking about `str_prop`, `atom_prop` and `intern_atom`. Those were previously calling the `atom` function which would check whether the given atom is known. In order to check this, the atom string must first be converted into an `Atom` such that it can be looked up in the `HashMap`. Do you want me to change the `atom` function such that it interns them with the XCB API directly instead of getting them from `self.atoms`? Or should I keep the `atom` function removed and intern them using XCB within each of those functions that take `&str` parameters? That way the caller can use arbitrary atoms instead of being limited by those known by `XcbConnection`.

sminez commented:
I am yeah. The issue with things always being looked up in the hash map is an oversight that does need fixing but not as part of this PR. Leave it for now and I'll get it sorted once your change is in 👍 This PR should only really be changing the hard coded `&str` instances to enums with calls to `into()` to convert them to `&str` .

There are a number of other things that need updating and fixing in this file but none of them should be done as part of this PR/issue 😄

Danacus committed:
gen_enum_with_slice macro

Danacus committed:
X atom types enum

Danacus commented:
I'm sorry for all the force pushing, I was struggling a bit with git and wanted to keep the commit history clean. I've made minimal changes to the `atom` function such that it still works and I removed `known_atom` because it has no uses anymore.

sminez commented:
@Danacus please stop force pushing changes in response to review comments: it makes things a lot harder to check in relation to the feedback that I am leaving. The git history can be squashed down for the merge if that is your concern.

You are still making changes to both the properties on the `XcbConnection` struct itself _and_ both altering method signatures and removing the `known_atom` method. I am not going to be able to merge this PR if it is doing anything more than what I have previously outlined before: introducing the enum and replacing hard coded strs with calls to `into()` on enum varients.

The diff should be really small.

It shouldn't be changing the types of the `atoms` or `auto_float_types` HashMaps and it should not be making any changes to methods (especially removing methods). As I said before: there are multiple things that need to be done to tidy up this file in the long run but those changes should not be made as part of this PR.

I need to be very careful about changes to `XcbConnection` and I want to make those changes myself so I can be happy that they have had sufficient manual testing: automated testing is very difficult to set up given that this involves interacting with the X server at this level.

Danacus commented:
> @Danacus please stop force pushing changes in response to review comments: it makes things a lot harder to check in relation to the feedback that I am leaving. The git history can be squashed down for the merge if that is your concern.

I didn't know that. I'll stop force pushing from now on. I does indeed make it harder to see recent changes, I'm sorry for that.

> It shouldn't be changing the types of the `atoms` or `auto_float_types` HashMaps and it should not be making any changes to methods (especially removing methods). As I said before: there are multiple things that need to be done to tidy up this file in the long run but those changes should not be made as part of this PR.

Oh okay, I did not understand that. Thanks for clearing things up!

> I need to be very careful about changes to `XcbConnection` and I want to make those changes myself so I can be happy that they have had sufficient manual testing: automated testing is very difficult to set up given that this involves interacting with the X server at this level.

I understand, I shouldn't be changing things in other people's projects that weren't requested, by bad. I'm used to just working on my own projects. I'm sorry if I wasted some of your time. Now I understand what exactly you are expecting from this PR.

Danacus committed:
Undo changes to HashMaps

Restored 'atoms' and 'auto_float_types' in XcbConnection.
Re-added known_atom method.
Changed uses of these HashMaps to accomodate for the changes.

sminez commented on src/core/xconnection.rs:
```diff
@@ -15,11 +15,60 @@ use crate::{
     Result,
 };
-use std::{cell::Cell, collections::HashMap};
+use std::{cell::Cell, collections::HashMap, convert::TryFrom, convert::TryInto};
```
`TryInto` isn't being used in this file is it?

Danacus commented on src/core/xconnection.rs:
```diff
@@ -930,8 +996,11 @@ impl XConn for XcbConnection {
-            Ok(s) => s.split("\0").any(|t| self.auto_float_types.contains(&t)),
+        match self.str_prop(id, Atom::NetWmWindowType.into()) {
+            Ok(s) => s.split("\0").any(|t| {
+                t.try_into()
```
`TryInto` is used here.
